### PR TITLE
Add Config and Migration plus review changes

### PR DESCRIPTION
**Pull request by aphraoh**
| Q             | A
| ------------- | ---
| Is bugfix?    | ❌
| New feature?  | ✔️
| Breaks BC?    | ✔️

### CODE REVIEW

**Comment by aphraoh** [comment]
Hi @vjik and @xepozz us requested except...

1. @xepozz Did not put "yiisoft/db-migration" to "suggested" section in composer.json as breaks build, please can we deal with this in another pull request.
2. @vjik not remove the empty di as this also cause problems (can not exactly remember what they were), please can we deal with this in another pull request.

File: config/di.php [hunk 1]
```diff
@@ -0,0 +1,6 @@
+<?php
+
+declare(strict_types=1);
+
+return [
```
Comment (samdark) [inline config/di.php]: Does it make sense to provide the config if it is empty?

File: src/Adapter.php [hunk 2]
```diff
@@ -6,17 +6,24 @@
...
 use Yiisoft\Db\Connection\ConnectionInterface;
 use Yiisoft\Db\Query\Query;
+use Yiisoft\Mutex\MutexFactoryInterface;
+use Yiisoft\Mutex\MutexInterface;
 
 final class Adapter implements AdapterInterface
 {
     /**
-     * @var int timeout
+     * @var MutexInterface mutex interface
```
Comment (samdark) [inline src/Adapter.php]: We're usually starting descriptions with capital letter and end these with dot. Not a huge issue though. Can fix it later.

```suggestion
     * @var MutexInterface Mutex interface.
```

File: src/Adapter.php [hunk 3]
```diff
@@ -92,7 +97,7 @@ public function push(MessageInterface $message): MessageInterface
 
     public function subscribe(callable $handlerCallback): void
     {
-        $this->runExisting($handlerCallback);
+        $this->run($handlerCallback, true, 5); // TWK TODO timeout should not be hard coded
```
Comment (samdark) [inline src/Adapter.php:102]: Do you plan to resolve it before merge?

File: src/Adapter.php [hunk 4]
```diff
@@ -115,10 +121,10 @@ public function withChannel(string $channel): self
      */
     protected function reserve(): array|null
     {
-        // TWK TODO ??? return $this->db->useMaster(function () {
-        // TWK TODO ??? if (!$this->mutex->acquire(__CLASS__ . $this->channel, $this->mutexTimeout)) {
-        // TWK TODO ??? throw new \Exception('Has not waited the lock.');
-        // TWK TODO ??? }
+        // TWK TODO what is useMaster in Yii3 return $this->db->useMaster(function () {
```
Comment (samdark) [inline src/Adapter.php:126]: @Tigrov, @terabytesoftw?

File: src/Adapter.php [hunk 5]
```diff
@@ -194,4 +200,28 @@ private function moveExpired(): void
...
+     *
+     * @param callable(MessageInterface): bool  $handlerCallback The handler which will handle messages. Returns false if it cannot continue handling messages
+     * @param bool $repeat whether to continue listening when queue is empty.
+     * @param non-negative-int $timeout number of seconds to sleep before next iteration.
+     */
+    public function run(callable $handlerCallback, bool $repeat, int $timeout = 0): void
+    { 
+        while ($this->loop->canContinue()) {
+            if ($payload = $this->reserve()) {
+                if ($handlerCallback(\unserialize($payload['job']))) {
```
Comment (samdark) [inline src/Adapter.php]: In general, we're not using PHP serialization for queue. Is it necessary?

Comment (xepozz) [inline src/Adapter.php]: Use https://github.com/yiisoft/queue/blob/master/src/Message/MessageSerializerInterface.php instead

**Comment by xepozz** [comment]
@aphraoh Do not remove discussions, do not use force-push, just add commits one by one, do not care about the rest.

As I wrote before, I don't like `yiisoft/db-migration` and `yiisoft/mutex` in `require` section. If the second is used in the adapter, then first is just declared one files with sql and it maybe useless. 

Move it to `suggests` composer section.

**Comment by samdark** [comment]
> just add commits one by one, do not care about the rest.

Yes. We take care about squashing commits when merging, but we need to see how it progressed while doing review. Also, force-pushing is quite confusing for GitHub UI.

File: src/Adapter.php (hunk 3 above)
Comment (aphraoh) [reply src/Adapter.php:102]: Can I do it after the merge, I believe the solution should be somewhere in the middleware but need to investigate a bit further so maybe that can be a lession for another day after the merge...

File: config/di.php (hunk 1 above)
Comment (aphraoh) [reply config/di.php]: Hi, have deleted it.

**Comment by aphraoh** [comment]
> > just add commits one by one, do not care about the rest.
> 
> Yes. We take care about squashing commits when merging, but we need to see how it progressed while doing review. Also, force-pushing is quite confusing for GitHub UI.

Oh dear, please send me a link to the process I should following if I would like to ammend a pull request in github :-|

File: src/Adapter.php (hunk 5 above)
Comment (aphraoh) [reply src/Adapter.php]: OK, will use this interface

**Comment by aphraoh** [comment]
Hi @xepozz @samdark @vjik , hopefully we can merge with the following outstanding...

1. // TWK TODO timeout should not be hard coded
2. // TWK TODO what is useMaster in Yii3
3. @xepozz, have removed yiisoft/db-migration from composer.json, but yiisoft/mutex still in required section cause used in Adopter (see constructor)

**Comment by aphraoh** [comment]
Apologies in advance if I have done any squeezed and forced commits, still googling to see exactly what you mean by that...

**Comment by samdark** [comment]
@aphraoh latest commits are fine.

**Review by samdark — APPROVED** [review]

**Comment by samdark** [comment]
Let's merge it and address comments separately. Good job, @aphraoh.

**Comment by aphraoh** [comment]
Sent from my iPhoneOn 3 Apr 2024, at 17:33, Alexander Makarov ***@***.***> wrote:﻿
Let's merge it and address comments separately. Good job, @aphraoh.

**Pull request merged**